### ISSUE TITLE
Unable to download Microsoft CSAF

### ISSUE THREAD

**Issue by Rafiot** [gocsaf/csaf#588]
I'm not sure what the issue is, but fetching the CSAF entries from Microsoft fails in a weird way: [downloader.log](https://github.com/user-attachments/files/17816521/downloader.log)

```
{"time":"2024-11-19T15:19:47+01:00","level":"INFO","msg":"AdvisoryFileProcessor.Process: \"https://msrc.microsoft.com/csaf/changes.csv\" has an invalid time stamp in line 89: parsing time \"2024-11-12T08:00:00.0000000\" as \"2006-01-02T15:04:05Z07:00\": cannot parse \"\" as \"Z07:00\""}
```

The call:

```
csaf_downloader -d csaf_microsoft https://msrc.microsoft.com/csaf/provider-metadata.json
```

Do you have any idea how to fix it?

**Comment by bernhardreiter** [comment]
A line from https://msrc.microsoft.com/csaf/changes.csv:

> 2024/msrc_cve-2024-49060.json,2024-11-15T08:00:00.0000000

Checking the format, this seems to be missing a timezone, so it maybe an invalid file.

https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#7113-requirement-13-changescsv has
> lines MUST be sorted by the current_release_date timestamp

https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#321122-document-property---tracking---current-release-date
> (current_release_date) with value type string with format date-time

And `date-time` is defined in https://json-schema.org/understanding-json-schema/reference/string as
> Dates and times are represented in RFC 3339, section 5.6. 

> "date-time": Date and time together, for example, 2018-11-13T20:20:39+00:00

https://datatracker.ietf.org/doc/html/rfc3339#section-5.6 then has an ABNF which
> time-offset     = "Z" / time-numoffset
> full-time       = partial-time time-offset

The time-offset is missing in the file. Best would be if the provider could fix this on their end.

**Comment by Rafiot** [comment]
I'm reporting it to Microsoft, and pointing them to your details, thanks!

**Comment by Rafiot** [comment]
Update on that: the time issue is fixed, but now I get this error:

```json
{"time":"2024-11-21T22:39:59+01:00","level":"WARN","msg":"Content type is not 'application/json'","url":"https://msrc.microsoft.com/csaf/2024/msrc_cve-2024-49028.json","content_type":"application/json; charset=utf-8"}
```

Their server returns `application/json; charset=utf-8` when you expect `application/json`, but both should be valid (?).

I couldn't find it in the specs, but I'll drop them another mail if needed.

**Comment by tschmidtb51** [comment]
> Their server returns `application/json; charset=utf-8` when you expect `application/json`, but both should be valid (?).

True. IMHO, both should be valid.
Flagging @bernhardreiter to suggest on how we want to deal with the situation in the checker in general.

**Comment by s-l-teichmann** [comment]
> > Their server returns `application/json; charset=utf-8` when you expect `application/json`, but both should be valid (?).
> 
> True. IMHO, both should be valid. Flagging @bernhardreiter to suggest on how we want to deal with the situation in the checker in general.

Yes, this should pass. Suggestion: If we detect an encoding other than UTF-8 we should issue a warning that the tools currently only supports UTF-8 documents. In the long run we may consider adding support for other encodings as well.

**Comment by bernhardreiter** [comment]
> If we detect an encoding other than UTF-8 we should issue a warning that the tools currently only supports UTF-8 documents. In the long run we may consider adding support for other encodings as well.

As far as I see it must be an ERROR and we MUST NOT support other encodings.

The standard https://datatracker.ietf.org/doc/html/rfc8259#section-8.1 is clear on this:
> JSON text exchanged between systems that are not part of a closed
   ecosystem MUST be encoded using UTF-8

For the `charset` information it has
> IANA Considerations
>
>  The media type for JSON text is application/json.

[..]
> Required parameters:  n/a
> Optional parameters:  n/a

The reported errata point https://www.rfc-editor.org/errata/eid5853 correctly argues that a `charset` MUST NOT be used. Also rfc6838's description of `application` does not have it.  I follow that interpretation and thus for the checker, I suggest we keep the error if we find a `charset=utf-8` parameter.

**Comment by Rafiot** [comment]
Alright, I'm talking to the Microsoft folks again. I don't really have a dog in this fight, I just don't want to implement my own downloader for fun.

**Comment by bernhardreiter** [comment]
> I'm talking to the Microsoft folks again.

Thanks a lot for reporting the problem to them!

> I don't really have a dog in this fight, I just don't want to implement my own downloader for fun.

Sounds like you do have an interest in a well implementable standard, though.  ;-)

**Comment by Rafiot** [comment]
Absolutely! And more importantly, compatible implementations of said standard :sob:

**Comment by adulau** [comment]
I'm really curious the origin of the issue has it seems the validation fails at Microsoft but also NCSC-NL too. 

https://github.com/cve-search/vulnerability-lookup/issues/94 

But maybe adding tests with known sources would be also interesting for the downloader?

**Comment by bernhardreiter** [comment]
>  it seems the validation fails at  [..] NCSC-NL 
> https://github.com/cve-search/vulnerability-lookup/issues/94

The failure is correct and the error message of the downloader shows that they've "invented" an 18th month. So it is not this not this issue.

**Comment by Rafiot** [comment]
I just sent a reminder to Microsoft so we hopefully get it sorted.

**Comment by Rafiot** [comment]
Alright they got back to me and will give me an answer by next week or so.

@bernhardreiter I have an alternative solution, let me know what you think: the downloader could accept the presence of the encoding, check if it is UTF-8 and raise the exception only if it is something else?

**Comment by bernhardreiter** [comment]
>  the downloader could accept the presence of the encoding, check if it is UTF-8 and raise the exception only if it is something else?

Just preparing my talk for the CSAF community days, and thus reading https://en.wikipedia.org/wiki/Robustness_principle#Criticism . So far I think we should be strict early in the CSAF 2.0 adoption.

**Comment by Rafiot** [comment]
I get your point, and I'll keep not having a dog in this fight. Depending on Microsoft feedback I'll or won't implement a dumb downloader that just fetches the files from their repo.

**Comment by Rafiot** [comment]
Just checking something, is that following the specs of not providing the encoding: https://github.com/gocsaf/csaf/blob/1daaed2c516d3fd674eb99c39dfc5f87ba43f78a/cmd/csaf_provider/controller.go#L177

**Comment by bernhardreiter** [comment]
> something, is that following the specs of not providing the encoding

Good catch, thanks for reporting!

**Comment by bernhardreiter** [comment]
The general question is now split out to #606. As the `charset=utf-8` should not be there, it would be very cool if Microsoft would fix this - no matter what the resolution of #606 is.

**Comment by Rafiot** [comment]
Alright, so I've been debugging a bit more and can you confirm that the warning with the encoding is **not** blocking? In the logs, it's a warning, so I don't think it is. 

What is blocking now (at least has an error in logs) is that the sha256 validation fails:

```{"time":"2024-12-17T21:19:26+01:00","level":"ERROR","msg":"Validation check failed","error":"SHA256 checksum of https://msrc.microsoft.com/csaf/2024/msrc_cve-2024-43644.json does not match"}```

And they don't provide the sha256 hash, only the sha512, but if I read your code properly, it shouldn't be a problem because the sha512 is there anyway. **But** their webserver returns a 2XX code instead of a 404 when querying a non-existing URL.
Could it be that the downloader gets the content of the response, treats it as the hash, and uses that to compare with the one it generated (and fails)?

**Comment by tschmidtb51** [comment]
> Alright, so I've been debugging a bit more and can you confirm that the warning with the encoding is **not** blocking? In the logs, it's a warning, so I don't think it is.

This is what I thought. (I wasn't really paying attention to the discussion as I was busy with the CSAF Community Days - I always wondered (and doubted) that the `charset=uft8` would prevent downloads but didn't found a quite minute to look into that.)

> **But** their webserver returns a 2XX code instead of a 404 when querying a non-existing URL.
> Could it be that the downloader gets the content of the response, treats it as the hash, and uses that to compare with the one it generated (and fails)?

Yes. This is my assumption from what I saw last night. So the actual problem preventing the download is that Microsoft delivers a login page to us (with HTTP status code 200) instead of reporting the "Not Found" 404.

**Comment by Rafiot** [comment]
It would probably also make sense to improve a bit this method there: https://github.com/gocsaf/csaf/blob/main/util/hash.go#L24

So it makes sure the string it gets is a hash (of the right size), and returns a less confusing error message.

**Comment by tschmidtb51** [comment]
Maybe that is in the works in #554

**Comment by Rafiot** [comment]
All good, that's implemented now.
